access/ms-teams/msapi/token.go

```diff
@@ -0,0 +1,129 @@
+package msapi
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"io"
+	"net/http"
+	"net/url"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/gravitational/teleport-plugins/lib/backoff"
+	"github.com/gravitational/trace"
+	"github.com/jonboulle/clockwork"
+)
+
+const (
+	getTokenBaseURL     = "https://login.microsoftonline.com"
+	getTokenContentType = "application/x-www-form-urlencoded"
+)
+
+// Token represents utility struct used for parsing GetToken resposne
+type Token struct {
+	AccessToken string `json:"access_token"`
+	ExpiresIn   int64  `json:"expires_in"`
+}
+
+// tokenWithTTL represents struct which handles token refresh on expiration
+type tokenWithTTL struct {
+	mu        sync.Mutex
+	token     Token
+	scope     string
+	expiresAt int64
+	baseURL   string
+}
+
+// Bearer returns current token value and refreshes it if token is expired.
+//
+// MS Graph API issues no refresh_token for client_credentials grant type. There also is no
+// extended validity window for this grant type. Hence, there is no way of preemprive refresh
+// the current token. We have to lock all requests and wait until the new token arrives.
+func (c *tokenWithTTL) Bearer(ctx context.Context, config Config) (string, error) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
```

Comment:
I would try to avoid doing any I/O (incl. network calls as getToken below does) under lock. Any way to refactor this do we're not making an API request while holding a lock?

Reply:
MS Graph API issues no `refresh_token` for `client_credentials` grant type. There also is no extended validity window for this grant type. Hence, there is no way of preemptive refresh the current token. We have to lock all requests and wait until the new token arrives.

I guess, that's my mistake, there's not much harm in doing multiple concurrent token refresh requests. Refactored.